packages/hardhat/contracts/LargeGrant.sol

```diff
@@ -0,0 +1,338 @@
+				index++;
+			}
+		}
+	}
+
+	function addGrant(
+		address _builder,
+		uint256 _grantId,
+		uint256[] memory _milestoneAmounts
+	) public onlyRole(OWNER_ROLE) {
+		GrantData storage grant = grants[_grantId];
+
+		if (grant.builder != address(0)) {
+			revert GrantAlreadyExists();
+		}
+
+		if (_builder == address(0)) {
+			revert ZeroAddress();
+		}
+
+		if (_milestoneAmounts.length == 0) {
+			revert NoMilestones();
+		}
+
+		grant.builder = _builder;
+
+		Stage storage stage = grant.stages.push();
+		stage.number = 1;
+
+		for (uint i = 0; i < _milestoneAmounts.length; i++) {
+			uint256 amount = _milestoneAmounts[i];
+			if (amount == 0) {
+				revert MilestoneZeroAmount();
+			}
+			stage.milestones.push(
+				Milestone({
+					number: uint8(i + 1),
+					amount: amount,
+					completed: false
+				})
+			);
+		}
+
+		emit AddGrant(_grantId, _builder, _milestoneAmounts);
+	}
+
+	function addGrantStage(
+		uint256 _grantId,
+		uint256[] memory _milestoneAmounts
+	) public onlyRole(OWNER_ROLE) {
+		GrantData storage grant = grants[_grantId];
+
+		if (grant.builder == address(0)) {
+			revert GrantDoesNotExist();
+		}
+
+		if (_milestoneAmounts.length == 0) {
+			revert NoMilestones();
+		}
+
+		Stage storage stage = grant.stages.push();
+
+		stage.number = uint8(grant.stages.length);
+
+		for (uint i = 0; i < _milestoneAmounts.length; i++) {
+			uint256 amount = _milestoneAmounts[i];
+			if (amount == 0) {
+				revert MilestoneZeroAmount();
+			}
+			stage.milestones.push(
+				Milestone({
+					number: uint8(i + 1),
+					amount: amount,
+					completed: false
+				})
+			);
+		}
+
+		emit AddGrantStage(
+			_grantId,
+			uint8(grant.stages.length),
+			_milestoneAmounts
+		);
+	}
+
+	function completeMilestone(
+		uint256 _grantId,
+		uint8 _stageNumber,
+		uint8 _milestoneNumber
+	) public whenNotPaused onlyRole(OWNER_ROLE) {
+		GrantData storage grant = grants[_grantId];
+
+		if (grant.builder == address(0)) {
+			revert GrantDoesNotExist();
+		}
+
+		if (_stageNumber > grant.stages.length) {
+			revert WrongStageNumber();
+		}
+
+		if (
+			_milestoneNumber > grant.stages[_stageNumber - 1].milestones.length
+		) {
+			revert WrongMilestoneNumber();
+		}
+
+		if (
+			grant
+				.stages[_stageNumber - 1]
+				.milestones[_milestoneNumber - 1]
+				.completed
+		) {
+			revert MilestoneAlreadyCompleted();
+		}
+
+		uint256 amount = grant
+			.stages[_stageNumber - 1]
+			.milestones[_milestoneNumber - 1]
+			.amount;
+
+		if (ERC20(tokenAddress).balanceOf(address(this)) < amount)
+			revert InsufficientContractFunds();
+
+		ERC20(tokenAddress).transfer(grant.builder, amount);
+
+		grant
+			.stages[_stageNumber - 1]
+			.milestones[_milestoneNumber - 1]
+			.completed = true;
+
+		emit CompleteMilestone(
+			_grantId,
+			_stageNumber,
+			_milestoneNumber,
+			amount
+		);
+	}
+
+	function completeMilestones(
+		uint256 _grantId,
+		uint8 _stageNumber,
+		uint8[] memory _milestoneNumbers
+	) public whenNotPaused onlyRole(OWNER_ROLE) {
```

Comment:
I think this will allow only particular grants milestones to be batch completed right? 

I think on admin view we will showing milestones from different grants but just cc @Pabl0cks once

Reply:
We decided to remove this 🙌